pkg/search/postgres/store_cache.go

```diff
@@ -408,6 +369,37 @@ func (c *cachedStore[T, PT]) GetIDsByQuery(ctx context.Context, query *v1.Query)
 	return c.underlyingStore.GetIDsByQuery(ctx, query)
 }
 
+func (c *cachedStore[T, PT]) getByQueryFn(ctx context.Context, query *v1.Query, fn func(obj PT) error) error {
+	if query == nil || query.EqualVT(search.EmptyQuery()) {
+		c.cacheLock.RLock()
+		defer c.cacheLock.RUnlock()
+		return c.walkCacheNoLock(ctx, fn)
+	}
+
+	identifiers, err := c.underlyingStore.GetIDsByQuery(ctx, query)
+	// Fallback to the underlying store on error.
+	if err != nil {
+		log.Errorf("Failed to get identifiers by query, falling back to get results by query: %v", err)
+		return c.underlyingStore.GetByQueryFn(ctx, query, fn)
+	}
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	for _, id := range identifiers {
+		obj, found := c.cache[id]
```

Comment:
Would it make sense to catch the context error here ?

Reply:
done and test added